### ISSUE TITLE
High CPU usage at idle and continuously looping fetch for extensions::setIcon

### ISSUE THREAD

**Issue by Infinitay** [web-scrobbler/web-scrobbler#3789]
**Describe the bug**
After a fresh restart of my PC and Chrome, I immediately noticed my computers fans spinning faster. I took at look at task manager and Chrome was the culprit. Taking a look at Chrome's task manager, The Web Scrobbler extension was using >= 100% of my CPU usage. At the time I opened a new tab and opened the developer console because at first it didn't explicitly say "Extension: Web Scrobbler", it only showed a service worker and pointed to chrome-extension:// with Web Scrobbler's ID (hhinaapppaileiechjoiifaancjggfjm). Immediately I noticed lots of logs to the console pointing to it's storage variable with settings and other data related to the extension. 

Furthermore, when expecting the service worker for the extension, I noticed there was an endless loop of a fetch call within `extensions::setIcon` @L10 within the `loadImagePathForServiceWorker` function.

**How to reproduce**
I'm not sure to be honest. I logged on to my PC for the first time today and relaunched my Chrome session for the previous night

**Expected behavior**
1. No high CPU usage
2. No endless fetch loop

**Screenshots**
![image](https://github.com/web-scrobbler/web-scrobbler/assets/6964154/dc90f86b-1844-48fc-8fe2-d6ba75c3517f)
![image](https://github.com/web-scrobbler/web-scrobbler/assets/6964154/06800a06-034d-4bd2-9307-2b3989b14673)
![image](https://github.com/web-scrobbler/web-scrobbler/assets/6964154/7ace28d4-0b76-455c-99d5-f229fa6cf0ae)

**Debug logs**
It's strongly recommended to grab [debug logs](https://github.com/web-scrobbler/web-scrobbler/wiki/Debug-the-extension) after reproducing the issue.

**Environment (please complete the following information)**
-   OS: Windows 10
-   Browser: Chrome, Version 113.0.5672.127 (Official Build) (64-bit)
-   Extension version: 3.0.5

**Additional context**

**Potentially resolved itself?**
As I was getting my environment data to submit this bug report, the endless loop of fetching the icon and calling of `setIcon` was stopped. This happened immediately after I went on the extensions page and right once I clicked on the "Accounts" tab. The odd thing is that I was already on the extension options page and clicked thru all of the option menus **except** for that Account tab. The moment I did my fans slowed down and when I saw Chrome's task manager the extension's CPU usage dropped.

I also just noticed that when switching tabs, those icon fetches gets called 8 times, twice really because the same icons are fetched twice.
![image](https://github.com/web-scrobbler/web-scrobbler/assets/6964154/0718a758-5181-4e14-aafe-b7e38810a9a7)
If I close a different tab with the context menu to not cause a tab switch, it fetches and calls `setIcon` 4 times. I suppose the reason why it happens 8 times is it fetches once per tab modification?

**Comment by yayuyokitano** [comment]
Yep you're right. With manifest v3 we can no longer rely on background script state as we did before, which has led to what is currently a very naive approach of just setting icon on every update.
I guess the fetch loop must have been related to the number of tabs somehow? But that also seems odd, normally content script shouldnt be loading until the page is refreshed.
I have a few ideas to improve the performance here in the future which would also render whatever happened here not a problem, though we have some high priority breaking bugs to fix first.
Thank you for reporting and showing that performance work is needed here!

**Comment by gigithegigiki** [comment]
> Yep you're right. With manifest v3 we can no longer rely on background script state as we did before, which has led to what is currently a very naive approach of just setting icon on every update. I guess the fetch loop must have been related to the number of tabs somehow? But that also seems odd, normally content script shouldnt be loading until the page is refreshed. I have a few ideas to improve the performance here in the future which would also render whatever happened here not a problem, though we have some high priority breaking bugs to fix first. Thank you for reporting and showing that performance work is needed here!

I just tried it, playing from spotify, today and **it crashed my PC** (memory over 4GB + CPU usage 99%)... 
I will not use it till you fix these serious bugs. 
I **cannot browse** in the internet **even if the scrobbler doesn't work**, it is lagging...
There is a **HUGE performance issue** in this version.
Will be patient till you fix it, but until then the app is not usable.

_Thanks for your effort_

**Comment by kaielvin** [comment]
Sorry if it is not related, but frequently getting 100 % CPU, logs showing:
`Error: Could not establish connection. Receiving end does not exist. at me (main.js:1:40990)`
every second.

**Comment by yayuyokitano** [comment]
The #3801 PR is probably causing the majority of this from my testing, will get this out relatively soon, and that might be enough to make performance no longer a major issue.

**Comment by yayuyokitano** [comment]
3.0.9 has now been pushed with major performance enhancements and is pending review. Please give the extension another whirl once you see 3.0.9 has been installed, and see if your issues persist, or if we have reached a point of the extension being usable.

**Comment by yayuyokitano** [comment]
Alright I will be closing this issue as it seems to no longer be a big problem. Will continue looking into it at a later date but pretty low priority.

The extension popup delay is completely unrelated to the size of the popup. I believe it is caused by the migration from a MV2 background page to an MV3 service worker, as this causes some changes to how the popup is loaded. This change was necessary and cannot be reverted, as chrome will soon stop accepting updates to extensions that use MV2 background pages. This was the primary reason for V3 update.

I will try some workarounds and see if I can make any progress on popup speed at a later time, though there are some other fixes such as broken connectors that have higher priority right now.

If you want to track this further, please make a new issue for it, since it is unrelated to this.
